content/subtleties.md

```diff
@@ -14,37 +14,40 @@ Also see Pierre Weis's [Frequently asked Questions about Caml](http://caml.inria
-  The signature says that the function will work with any type, but the compiler
-  thinks that your function might only work with a single input and output type
-  (`int`, in the first error).  The value restriction is the name of the heuristic
-  that the compiler uses to make this decision.  For more information see the section on
-  [Side effects and weak polymorphism](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#side-effects-and-weak-polymorphism)
-  in *Real World OCaml*,
-  [Section 5.1 Weak polymorphism and mutation](http://caml.inria.fr/pub/docs/manual-ocaml/polymorphism.html#sec51) 
-  in the OCaml Manual, [Common Error Messages](http://ocaml.org/learn/tutorials/common_error_messages.html#The-type-of-this-expression-contains-type-variables-that-cannot-be-generalized) at ocaml.org, and ["My function is not polymorphic (enough) ?"](http://caml.inria.fr/pub/old_caml_site/FAQ/FAQ_EXPERT-eng.html#eta_expansion) and subsequent sections in "Frequently asked Questions about Caml".
+  In OCaml and other ML-style languages that support mutability, only _values_ can have polymorphic types, hence the name -- "value restriction". The underlying reason, is that if an expression is mutable (an since we don't know its type, we must assume that it is mutable), then it could be assigned multiple times, each time with a value of a different type, thus we can't guarantee that the type of a value would be preserved. That will eventually break the type soundness. Originally, values were defined as syntatic constants. I.e., something that is a static constant on the source level, is considered a value, e.g., `1`, `"hello"`, `None`, `fun x -> x`. However, in OCaml this restriction is relaxed, and besides syntactic constants OCaml allows some other classes of expressions to be generalized (i.e., to have the polymoprhic type). First of all, OCaml allows polymorphism for _non-expansive_ expressions, where an expression is non-expansive if it doesn't have any observable side-effects during the evaluation. That basically extends the notion of constant from purely syntactic definition (something that looks like a constant, is a constant) to semantic definition (something that acts like a constant, is a constant). For example, this restiction allows the following expressions to have polymorphic type: `lazy None`, ` let _ = ref None in 1`, `assert false`. But that's not all, OCaml is able to prove that it is safe to generalize certain type variables, by performing their variance analysis. It was [observed]((https://caml.inria.fr/pub/papers/garrigue-value_restriction-fiwflp04.pdf)) that there is a nice interaction between subtyping, variance, and value restriction. If a value occurs only in covariant positions, then it can be safely generalized and given the polymorphic type. With this rule, it is possible to give the polymorphic type to the following expressions: `fun () -> []`, `fun () -> ref None`. 
```

Comment:
My first non-substantive comment: some paragraph breaks in here would be useful. It's too big a block of text to read at once. I'd insert a couple of breaks.

My second non-substantive comment: although there's no technical restriction on the length of lines in markdown files, I find it is much easier to read them during editing and maintenance if they're only 80 columns wide or less.

Reply:
The first issue is addressed.

Concerning the line length, on a long run it is better to allow markdown to perform the renderning and write all paragraphs as one big lines. It works better with merging and diffing, and other markdown constructs, including paragraphing. 

Of course, it is ugly, but in emacs we have plenty of options for line-wrapping. I believe, that vi is not significantly behind :)